### PR TITLE
Review topk for shape inference aspects

### PR DESCRIPTION
**Pull request by praasz**
### Details:
- label and dimension propagation
- partial value and label propagation
- preserve partial value and labels
- add evaluate upper, lower and label
- introduce shape_infer which return inferred shapes as return.
- add new `get_input_const_data_as_shape` which allow validate values by element type and pre-process it.

### Tickets:
 - 97259
 - 97279
 - 97328
 - 97377
 - 97426
 - 97480
 - 97529
 - 97578

### CODE REVIEW

File: src/core/shape_inference/include/shape_infer_transformations.hpp
Comment (tomdol) [inline src/core/shape_inference/include/shape_infer_transformations.hpp:53]: Could the `static_cast<T>` be used here instead? I don't see the benefit of creating an empty temporary object without any state.

File: src/core/shape_inference/include/shape_infer_transformations.hpp
Comment (tomdol) [inline src/core/shape_inference/include/shape_infer_transformations.hpp:39]: This member seems to be unused.

File: src/core/shape_inference/include/topk_shape_inference.hpp
Comment (tomdol) [inline src/core/shape_inference/include/topk_shape_inference.hpp:31]: Since the `Cast` structure doesn't perform any validation of the types or the value of `k` I think using `static_cast<T>` could be used here directly.

File: src/core/shape_inference/include/topk_shape_inference.hpp
Comment (tomdol) [inline src/core/shape_inference/include/topk_shape_inference.hpp:18]: I'd rather see this as a free function or a lambda. Is there any reason to wrap this in a struct?

File: src/core/shape_inference/include/utils.hpp
Comment (tomdol) [inline src/core/shape_inference/include/utils.hpp:73]:
```suggestion
        std::transform(static_cast<const dtype*>(ptr), static_cast<const dtype*>(ptr) + size, out_it, std::forward<UnaryOperation>(func));
```

And similarily for all std::transform calls below.

File: src/core/shape_inference/include/utils.hpp
Comment (tomdol) [inline src/core/shape_inference/include/utils.hpp:130]: std::forward the functor here (and other locations below)

**Review by tomdol — CHANGES_REQUESTED** [review]

File: src/core/shape_inference/include/topk_shape_inference.hpp
Comment (praasz) [reply src/core/shape_inference/include/topk_shape_inference.hpp:18]: Yes, to pass it as functor in `get_input_const_data_as_shape` and correct deduce type in operator() for input which got fundamental type of tensor data element.
Free function used here as template need to be specified and one "result type" is know as `TDimValue` but 2nd is runtime element type in tensor.
Lambda not support `auto` and if I use e.g. `TDimValue` it will be `size_t` or `int64_t`, then k value validation can give false results.

Assume `k` tensor is `i64` and we do inference for static shape so dimension value type is `size_t`
then for values -1,-2, etc. the tensor value will be casted to `size_t` which give positive number and always pass the check
or when `size_t` is 32bit then values greater than (int64_max/2) will be casted to 0..u32.max instead be rejected as too large.

File: src/core/shape_inference/include/shape_infer_transformations.hpp
Comment (praasz) [reply src/core/shape_inference/include/shape_infer_transformations.hpp]: Done

File: src/core/shape_inference/include/shape_infer_transformations.hpp
Comment (praasz) [reply src/core/shape_inference/include/shape_infer_transformations.hpp]: Removed.

File: src/core/shape_inference/include/utils.hpp
Comment (praasz) [reply src/core/shape_inference/include/utils.hpp]: Done

File: src/core/shape_inference/include/utils.hpp
Comment (praasz) [reply src/core/shape_inference/include/utils.hpp]: Done

File: src/core/shape_inference/include/utils.hpp
Comment (praasz) [reply src/core/shape_inference/include/utils.hpp]: Done

File: src/core/shape_inference/include/topk_shape_inference.hpp
Comment (praasz) [reply src/core/shape_inference/include/topk_shape_inference.hpp]: Done.

**Review by tomdol — APPROVED** [review]

**Comment by mlukasze** [comment]
@jane-intel || @mitruska could you take a look, please?

File: src/core/src/op/topk.cpp
Comment (mitruska) [inline src/core/src/op/topk.cpp:384]: Let's say there is `ParialShape{{2, 12}, {3, 10}, {4, 11}}` with overlapping interval dimension and labels {a, b, c}.
The final compatible static shape can be `Shape{4, 5, 6}` as well as `Shape{6, 5, 4}` or `Shape{5, 6, 4}` 
(...and other combinations).

The values taken as lower are: `{2, 3, 4}`, and upper `{12, 11, 10}` 
so the corresponding indices values for max top3 are:
indices lb: {2, 1, 0} => labels {c, b, a}
indices ub: {0, 2, 1} => labels {a, c, b}

for Shape{5, 6, 4} max top3 indices are {1, 0, 2} and the labels should be {b, a, c}.

So the proposed approach seems to be not universal for labels evaluation.

**Review by mitruska — COMMENTED** [review]
The implementation of `evaluate_lower/upper/label` needs to be discussed and verified.
Details in the comments.

*For the files shared between PRs I've left my suggestions in https://github.com/openvinotoolkit/openvino/pull/15004

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp:384]: Yes that is true, as I've used approach takes uppers and lowers and sort separately but it can result with illegal dimension or when static values will be used order will not be correct.

So maybe use other approach like:
As upper bound is usually more important so do evaluate for uppers only and then take their lowers and labels, and sort it that way. 
But if there is any overlapping dimension in this set then make output dynamic shape as order cannot be defined until static values will be used.

Example:
PartialShape{{2,3}, {10,12}, {1,4}, {5, 6}}
k = 4, max -> dynamic as there is overlap between {2,3}, and {1,4}
k = 3, max -> dynamic or there is overlap and indices cannot be specified for 3rd dimension
k = 2, max -> {{10, 12}, {5,6}} no overlaps
k = 4, min -> dynamic
k  = 3, min -> dynamic
k = 1, min -> dynamic 

Then:
Evaluate upper will do evaluation for uppers bounds but check corresponding lower for overlap if there is no over lap return true, and output tensor will be filled with uppers.

Evaluate lower will also do evaluation for uppers but fill with lower for corresponding topk uppers and return true if there is no overlap


@jane-intel @mitruska Share opinion about this solution or should I do it in another way.

**Comment by praasz** [comment]
> The implementation of `evaluate_lower/upper/label` needs to be discussed and verified. Details in the comments.
> 
> *For the files shared between PRs I've left my suggestions in #15004

So I will update these shared files in #15004

File: src/core/tests/type_prop/top_k.cpp
Comment (praasz) [reply src/core/tests/type_prop/top_k.cpp]: I've change implementation for bounds evaluation and in this case returned shape will got dynamic dimensions and no labels as is not possible to set correct order until static data will be provided.

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp]: Yes, true.
Change implementation and if there is no overlaps between partial values then order can be correctly calculated and default evaluator will work.

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp]: Implementation changed.

**Comment by mlukasze** [comment]
@jane-intel should we wait for you or it's good to merge?

File: src/core/src/op/topk.cpp
Comment (mitruska) [inline src/core/src/op/topk.cpp:135]: As the dimension bounds are inclusive, the case with equal bounds is still an overlap.
Also there is a note in the [TopK-3](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/sort/TopK_3.md) specification:
"If there are several elements with the same value then their output order is not determined."
```suggestion
               auto no_overlap = (is_max_mode && ov::cmp::lt(ub, prev_bound)) ||
                                 (!is_max_mode && ov::cmp::lt(prev_bound, lowers[idx]));
```

File: src/core/src/op/topk.cpp
Comment (mitruska) [reply src/core/src/op/topk.cpp]: The solution to return `false` for overlapping dimensions sounds better to me.
But in general the TopK in ShapeOf subgraph seems to be kind of theoretical case.

@jane-intel Could you confirm that the implementation of evaluate `evaluate_lower/upper/label` is applicable for TopK operation?

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp:135]: No, the lower then will not give correct results, there can be several dimension with same size (static size) and there will be no overlap ( will be sorted as reference implementation do)

**Review by mitruska — APPROVED** [review]
Approved, but the need of evaluate_lower/upper/label implementation for TopK should be confirmed before merge.

**Comment by mlukasze** [comment]
@jane-intel looks like ready to merge, could you please double check it?

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp]: As discussed off-line the bound evaluation has been removed, and after check if required will be provide in separate PR.
Will update it if required in new PR.

File: src/core/src/op/topk.cpp
Comment (praasz) [reply src/core/src/op/topk.cpp]: Right now removed it from this PR and if bound evaluation required will provided as other PR.

**Review by jane-intel — APPROVED** [review]

**Pull request merged**
